File: frontend/src/api/enums.ts
import { z } from 'zod'

import type { components } from './generated/schema'

// 生成 union（OpenAPI 由来）を型の真実源とし、実行時値は as const 配列で持つ。
// 混入は satisfies、欠落は Assert<false になる条件型> で双方向に検査する。
type Assert<T extends true> = T

export type CustomerStatus = components['schemas']['CustomerStatus']
export type CustomerArea = components['schemas']['CustomerArea']
export type UserRole = components['schemas']['UserRole']
export type ActivityType = components['schemas']['ActivityType']
export type VisitStatus = components['schemas']['VisitStatus']

export const CUSTOMER_STATUS = [
  'prospect',
  'negotiating',
  'won',
  'lost',
  'dormant',
] as const satisfies readonly CustomerStatus[]
export type _CustomerStatusExhaustive = Assert<
  [CustomerStatus] extends [(typeof CUSTOMER_STATUS)[number]] ? true : false
>

export const CUSTOMER_AREA = [
  'tokyo',
  'kanagawa',
  'saitama',
  'chiba',
  'other',
] as const satisfies readonly CustomerArea[]
export type _CustomerAreaExhaustive = Assert<
  [CustomerArea] extends [(typeof CUSTOMER_AREA)[number]] ? true : false
>

export const ACTIVITY_TYPE = [
  'visit',
  'call',
  'email',
  'online',
] as const satisfies readonly ActivityType[]
export type _ActivityTypeExhaustive = Assert<
  [ActivityType] extends [(typeof ACTIVITY_TYPE)[number]] ? true : false
>

export const VISIT_STATUS = [
  'planned',
  'done',
  'cancelled',
] as const satisfies readonly VisitStatus[]
export type _VisitStatusExhaustive = Assert<
  [VisitStatus] extends [(typeof VISIT_STATUS)[number]] ? true : false
>

export const USER_ROLE = ['sales', 'manager'] as const satisfies readonly UserRole[]
export type _UserRoleExhaustive = Assert<
  [UserRole] extends [(typeof USER_ROLE)[number]] ? true : false
>

export const customerStatusSchema = z.enum(CUSTOMER_STATUS)
export const customerAreaSchema = z.enum(CUSTOMER_AREA)
export const activityTypeSchema = z.enum(ACTIVITY_TYPE)
export const visitStatusSchema = z.enum(VISIT_STATUS)

// 表示ラベル（内部キーは英小文字、表示は日本語。仕様）
export const customerStatusLabels: Record<CustomerStatus, string> = {
  prospect: '見込み',
  negotiating: '商談中',
  won: '受注',
  lost: '失注',
  dormant: '休眠',
}

export const customerAreaLabels: Record<CustomerArea, string> = {
  tokyo: '東京',
  kanagawa: '神奈川',
  saitama: '埼玉',
  chiba: '千葉',
  other: 'その他',
}

export const activityTypeLabels: Record<ActivityType, string> = {
  visit: '訪問',
  call: '電話',
  email: 'メール',
  online: 'オンライン会議',
}

export const visitStatusLabels: Record<VisitStatus, string> = {
  planned: '予定',
  done: '完了',
  cancelled: 'キャンセル',
}

export const userRoleLabels: Record<UserRole, string> = {
  sales: '営業',
  manager: 'マネージャー',
}

File: backend/app/enums.py
import enum


class UserRole(str, enum.Enum):
    sales = "sales"
    manager = "manager"


class CustomerStatus(str, enum.Enum):
    prospect = "prospect"
    negotiating = "negotiating"
    won = "won"
    lost = "lost"
    dormant = "dormant"


class CustomerArea(str, enum.Enum):
    tokyo = "tokyo"
    kanagawa = "kanagawa"
    saitama = "saitama"
    chiba = "chiba"
    other = "other"


class ActivityType(str, enum.Enum):
    visit = "visit"
    call = "call"
    email = "email"
    online = "online"


class VisitStatus(str, enum.Enum):
    planned = "planned"
    done = "done"
    cancelled = "cancelled"


class AgentWorkflowType(str, enum.Enum):
    meeting_prep = "meeting_prep"
    risk_review = "risk_review"
    follow_up = "follow_up"


class AgentRunStatus(str, enum.Enum):
    pending = "pending"
    running = "running"
    waiting_for_approval = "waiting_for_approval"
    completed = "completed"
    failed = "failed"
    cancelled = "cancelled"


class AgentStepStatus(str, enum.Enum):
    running = "running"
    completed = "completed"
    failed = "failed"


class AgentApprovalStatus(str, enum.Enum):
    pending = "pending"
    approved = "approved"
    edited_and_approved = "edited_and_approved"
    rejected = "rejected"
    persisted = "persisted"
    persist_failed = "persist_failed"
    expired = "expired"
    cancelled = "cancelled"


class AgentActionType(str, enum.Enum):
    activity_log = "activity_log"
    task = "task"
    memo = "memo"
    email_draft = "email_draft"


class AgentIdempotencyStatus(str, enum.Enum):
    in_progress = "in_progress"
    succeeded = "succeeded"
    failed = "failed"


class AgentIdempotencyFailureKind(str, enum.Enum):
    retryable_before_side_effect = "retryable_before_side_effect"
    permanent = "permanent"
    unknown_after_possible_side_effect = "unknown_after_possible_side_effect"


class KnowledgeVisibility(str, enum.Enum):
    all_sales = "all_sales"
    managers_only = "managers_only"
    owner_team = "owner_team"
    customer_scoped = "customer_scoped"
    private = "private"
